### PR TITLE
[RouterDebugCommand] add link to Controllers

### PR DESCRIPTION
**Pull request by nicoweb**
| Q             | A
| ------------- | ---
| Branch?       | master for features
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Adds a link to the controller method on your IDE from dev's terminal: 

pr-debug-router

Configuration in your `services.yaml`:

```yaml
parameters:
    debug.file_link_format: phpstorm://open?file=%%f&line=%%l
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 1]
```diff
@@ -47,16 +55,19 @@ protected function describeRouteCollection(RouteCollection $routes, array $optio
 
         $tableRows = [];
         foreach ($routes->all() as $name => $route) {
+            $controller = $route->getDefault('_controller');
+            list ($controllerClass, $controllerMethod) = explode('::', $controller);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: can be done inside transformControllerToLink - should be more robust: not all controllers have `::` inside - eg invokable ones - and controllers can be arrays, or even regular callables.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 2]
```diff
@@ -448,6 +457,31 @@ private function formatRouterConfig(array $config): string
...
+            } elseif (false !== strpos($controller, '::')) {
+                $r = new \ReflectionMethod($controller);
+            } else {
+                $r = new \ReflectionFunction($controller);
+            }
+        } catch (\ReflectionException $e) {
+            return $anchorText;
+        }
+
+        $fileLink = $this->fileLinkFormatter->format($r->getFileName(), $r->getStartLine());
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: `$this->fileLinkFormatter` can be null

**Comment by fabpot** [comment]
@nicoweb Can you have a look at the comments and fix this PR accordingly? Thank you.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 2 above)
Comment (nicoweb) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: thanks @chalasr for your feedback, I updated.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml [hunk 3]
```diff
@@ -111,6 +111,7 @@
 
         <service id="console.command.router_debug" class="Symfony\Bundle\FrameworkBundle\Command\RouterDebugCommand">
             <argument type="service" id="router" />
+            <argument type="service" id="debug.file_link_formatter" on-invalid="null" />
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml]: I added `on-invalid="null"` for consistency.

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicoweb.

**Pull request merged**